java/query/GraqlMatch.java

```diff
@@ -81,22 +80,91 @@
         sortVarsArInScope();
 
-        this.hash = Objects.hash(this.conjunction, this.filter, this.sorting, this.offset, this.limit);
+        this.hash = Objects.hash(this.conjunction, this.modifier);
+    }
+
+    public static class Modifier {
+
+        private final List<UnboundVariable> filter;
+        private final Sortable.Sorting sorting;
+        private final Long offset;
+        private final Long limit;
+
+        private final int hash;
+
+        public Modifier(List<UnboundVariable> filter, @Nullable Sortable.Sorting sorting, @Nullable Long offset,
+                        @Nullable Long limit) {
+            this.filter = list(filter);
+            this.sorting = sorting;
+            this.offset = offset;
+            this.limit = limit;
+            this.hash = Objects.hash(this.filter, this.sorting, this.offset, this.limit);
+        }
+
+        public List<UnboundVariable> filter() {
+            return filter;
+        }
+
+        public Optional<Long> offset() {
+            return Optional.ofNullable(offset);
+        }
+
+        public Optional<Long> limit() {
+            return Optional.ofNullable(limit);
+        }
+
+        public Optional<Sortable.Sorting> sort() {
+            return Optional.ofNullable(sorting);
+        }
+
+        public boolean isEmpty() {
+            return filter.isEmpty() && sorting == null && offset == null && limit == null;
+        }
+
+        @Override
+        public String toString() {
+            StringBuilder syntax = new StringBuilder();
+            if (!filter.isEmpty()) {
+                syntax.append(GET);
+                final String varsStr = filter.stream().map(UnboundVariable::toString).collect(joining(COMMA_SPACE.toString()));
+                syntax.append(SPACE).append(varsStr);
+                syntax.append(SEMICOLON);
+            }
+            if (sorting != null) syntax.append(SORT).append(SPACE).append(sorting).append(SEMICOLON).append(SPACE);
+            if (offset != null) syntax.append(OFFSET).append(SPACE).append(offset).append(SEMICOLON).append(SPACE);
+            if (limit != null) syntax.append(LIMIT).append(SPACE).append(limit).append(SEMICOLON).append(SPACE);
+            return syntax.toString();
+        }
+
+        @Override
+        public boolean equals(final Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            final Modifier modifier = (Modifier) o;
+            return Objects.equals(filter, modifier.filter) && Objects.equals(sorting, modifier.sorting)
+                    && Objects.equals(offset, modifier.offset) && Objects.equals(limit, modifier.limit);
+        }
+
+        @Override
+        public int hashCode() {
+            return hash;
+        }
+    }
+
+    private List<UnboundVariable> retrievedVars() {
+        if (modifier.filter.isEmpty()) return variablesNamedUnbound;
+        else return modifier.filter;
```

Comment:
in the internal query validation, we used to first check filter vars if they exist, otherwise use the named vars. This is exposed as a new match-level method `retrievedVars()`